ykrt/src/mt.rs

```diff
@@ -219,215 +217,206 @@ impl MTThread {
+                    if self.inner.tracing.is_none() {
                         return None;
-                    } else {
-                        // ...and we started at this Location, so we've got a complete loop!
                     }
-                } else {
-                    // Another thread is tracing this location.
-                    if Arc::strong_count(&loc_tid) == 1 {
-                        // The other thread has stopped. Mark this Location as DONT_TRACE.
-                        loc.store(State::phase_dont_trace(), Ordering::Relaxed);
-                        return None;
-                    } else {
-                        // The other thread is still alive.
-                        loc.store(lp, Ordering::Relaxed);
-                        mem::forget(loc_tid);
-                        return None;
+                    match loc.lock() {
+                        Ok(x) => ls = x,
+                        Err(()) => unreachable!(),
                     }
                 }
-
-                // Stop the tracer
-                let sir_trace = Rc::get_mut(&mut self.inner)
-                    .unwrap()
-                    .tracer
-                    .take()
-                    .unwrap()
-                    .0
-                    .stop_tracing()
-                    .unwrap();
-
-                // Start a compilation thread.
-                let mtx = Arc::new(Mutex::new(None));
-                let mtx_cl = Arc::clone(&mtx);
-                thread::spawn(move || {
-                    let compiled = compile_trace::<I>(sir_trace).unwrap();
-                    *mtx_cl.lock().unwrap() = Some(Box::new(compiled));
-                    // FIXME: although we've now put the compiled trace into the mutex, there's no
-                    // guarantee that the Location for which we're compiling will ever be executed
-                    // again. In such a case, the memory has, in essence, leaked.
-                });
-                let new_state = State::phase_compiling(mtx);
-                loc.store(new_state, Ordering::Release);
-
-                return None;
             }
-            PHASE_COMPILING => {
-                // We need to free the memory allocated earlier. To ensure we don't race with
-                // another thread and both try and free the memory, we need to transition to
-                // PHASE_LOCKED so that only this thread will try to free the memory.
-                if loc.compare_and_swap(lp, State::phase_locked(), Ordering::Acquire) != lp {
-                    // We raced with another thread that's probably transitioned this Location to
-                    // PHASE_LOCK.
-                    return None;
+            let hl = unsafe { ls.hot_location() };
+            let hl_ptr = hl as *mut _ as *mut ();
+            match hl {
+                HotLocation::Compiled(tr) => {
+                    // FIXME: we need to refcount the compiled trace to know how many threads are
```

Comment:
Why to we worry about how many threads are executing a trace? Currently we don't discard trace code.

Reply:
In the long run we may want to recompile code and at the moment we can't free a trace because we can never be sure if someone's executing in it. This "FIXME" probably won't be addressed for a long time!

Reply:
Can we qualify the statement?

> // FIXME: If we ever decide to evict/invalidate trace code, then ...`

Reply:
Good point. Fixed in a3d9eb55c15524789ca14428ab41a73114be3e7e.